src/osd/PGLog.cc

```diff
@@ -1146,37 +1146,41 @@ namespace {
       on_disk_can_rollback_to = info.last_update;
       missing.may_include_deletes = false;
 
-      return seastar::repeat([ch=std::move(ch),
```

Comment:
@zhscn could explain a little bit on which variable(s) are used after being freed?

Reply:
The variable "start" can be used after being freed, at line [1160](https://github.com/ceph/ceph/pull/48505/files#diff-b293fb673637ea53b5874bbb04f8f0638ca39cab009610e2cbc40a867bca4906L1160)

Reply:
but `start` is in the capture list of the lambda passed to `seastar::repeat()`, the lambda in question is kept alive until the continuation of `seastar::repeat()` completes. i think the variables captured by it should have the same life cycle of the lambda. am i right?

Reply:
> the variables captured by it should have the same life cycle of the lambda. 

Yes, but the repeat body might be moved to somewhere else in `seastar::repeat`:
```cpp
future<> repeat(AsyncAction&& action) noexcept {
    using futurator = futurize<std::invoke_result_t<AsyncAction>>;
    static_assert(std::is_same<future<stop_iteration>, typename futurator::type>::value, "bad AsyncAction signature");
    for (;;) {
        // Do not type-erase here in case this is a short repeat()
        auto f = futurator::invoke(action);

        if (!f.available() || f.failed() || need_preempt()) {
            return [&] () noexcept {
                memory::scoped_critical_alloc_section _;
                auto repeater = new internal::repeater<AsyncAction>(std::move(action)); // <- the lambda is moved
                auto ret = repeater->get_future();
                internal::set_callback(std::move(f), repeater);
                return ret;
            }();
        }

        if (f.get0() == stop_iteration::yes) {
            return make_ready_future<>();
        }
    }
}
```
When the body is moved, in the callback lambda, the captured reference of `start` still references the original repeat body, this causes use after moved.

We can write a test case:
```cpp
#include <iostream>
#include <optional>
#include <string>

void foo(auto &s) { s = "foo"; }

int main() {
  auto f1 = [s = std::make_optional<std::string>()]() mutable {
    foo(s);
    std::cout << *s << '\n';
    return [&s] { std::cout << *s << '\n'; }; // this output is empty after the f1 moved
  };
  auto callback = f1();
  auto f2 = std::move(f1);
  callback();
}
```

Reply:
thanks for the detailed explanation. i tested the test with https://godbolt.org/z/1KeadPvf7. `foo\n` is always printed. did i miss anything?

Reply:
the output of `callback()` is empty in this case. If the refernce is still valid after `f1` is moved, we'll get `foo\nfoo\n`.

A more detailed code snippet: https://godbolt.org/z/KPKfedd1c

Reply:
ahh, thank you very much! now i see. so the variable pointed by the captured reference in the inner lambda is long gone along with the outer lambda after the original outer lambda is moved away when constructing `internal::repeater`.

Reply:
we might need to audit the code to see if we are misusing `seastar::repeat()` for constructing variables using the capture list of the lambda passed to it..